src/import/duplicate_detect.rs

```diff
@@ -67,48 +67,42 @@ impl<S: Snapshot> DuplicateDetector<S> {
         Ok(Some(ret))
     }
 
-    fn read_next_user_key(&mut self, duplicate_pairs: &mut Vec<KvPair>) -> Result<()> {
-        let start_key = self.iter.key();
-        let (current_key, commit_ts) = Key::split_on_ts_for(start_key)?;
-        if commit_ts <= self.min_commit_ts {
+    fn move_to_next_import_key(&mut self) -> Result<Option<(Vec<u8>, TimeStamp)>> {
+        while self.iter.valid().map_err(from_kv_error)? {
+            let (current_key, commit_ts) = Key::split_on_ts_for(self.iter.key())?;
+            if commit_ts > self.min_commit_ts {
+                return Ok(Some((current_key.to_vec(), commit_ts)));
+            }
             self.iter.next().map_err(from_kv_error)?;
-            return Ok(());
         }
+        Ok(None)
+    }
+
+    fn collect_current_key_duplicate(
+        &mut self,
+        start_key: Vec<u8>,
+        end_commit_ts: TimeStamp,
+        duplicate_pairs: &mut Vec<KvPair>,
+    ) -> Result<()> {
         let write = WriteRef::parse(self.iter.value()).map_err(from_txn_types_error)?;
-        match write.write_type {
+        let mut write_info = match write.write_type {
```

Comment:
I'm afraid you may need to check the gc_fence to ignore the invalidated write records.

Reply:
No, it is an import key.